[TranHieuNghia] iP

rockman007372 opened the pull request:
# DUKE - Task Management Bot

> "Don't trust anything you read on the internet" - Lee Kuan Yew

DUKE frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~fast~ **super fast** to use!

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/admin/ip-w4.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you

And it is **FREE**!

Features:

- [ ] Managing tasks
- [ ] Managing deadlines 
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

rockman007372 committed:
Add greet, echo and exit features

rockman007372 committed:
Rename bot and separate logics into functions

rockman007372 committed:
Add Add and List funtions and rename bot

rockman007372 committed:
Add Task class

rockman007372 committed:
Add Mark as Done feature

rockman007372 committed:
Add Todos, Events, Deadlines class

rockman007372 committed:
Add exceptions

rockman007372 committed:
Add NoTaskException

rockman007372 committed:
Modify listTask() to include edgecase

rockman007372 committed:
Add TaskOutOfBoundException

rockman007372 committed:
Rename class to Duke, implement Enum and switch cases, overhaul of most Duke's methods

rockman007372 committed:
Invert isValidTask logic

rockman007372 committed:
Add Level-6

rockman007372 committed:
Add enum

rockman007372 committed:
Change bot messages to be more civil

rockman007372 committed:
"Add jUnit and fix logic bug in checkValidity"

rockman007372 committed:
Change constant name to SCREAMING_SNAKE font

rockman007372 committed:
Modify output of the bot

rockman007372 committed:
Add save and load features

rockman007372 committed:
Overhaul bot text system

rockman007372 committed:
Change String to LocalDate in Task subclasses

rockman007372 committed:
Add LocalDate

rockman007372 committed:
Merge branch 'branch-Level-7'

Add features to load and save task list

Once the program restarts, the task list is reset.

The users may want to save their task list for the next session.

Let's create a save file for the task list and load it when the program restarts.

rockman007372 committed:
Revert accidental change in Event class

rockman007372 committed:
Resolve merge conflict between Level 7 and Level 8

rockman007372 committed:
Refractor code completely to add OOP

rockman007372 committed:
Organise classes into packages

rockman007372 committed:
Fix typos and remove excess import statements

rockman007372 committed:
Refractor into duke package

rockman007372 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Merge with gradle branch

rockman007372 committed:
Add Gradle

rockman007372 committed:
Add a more general addTask method in TaskList

rockman007372 committed:
Add 2 test cases in TaskListTest and add equals method in TaskList

rockman007372 committed:
Create JAR file

rockman007372 committed:
Add Javadoc to half of public methods

rockman007372 committed:
Fix coding standard for switch statement

rockman007372 committed:
Add find command

rockman007372 committed:
Delete a javadoc comment to merge branch again

rockman007372 committed:
Change some javadoc to do merge again

rockman007372 committed:
Add javadoc

rockman007372 committed:
Merge branch 'branch-A-CodingStandard'

Refractor the code to follow Coding Standard

rockman007372 committed:
Add find command

lhy-hoyin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+                String fullCommand = ui.readCommand();
+                Command c = Parser.parse(fullCommand);
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
```
I like how this is very clean. Good job!

lhy-hoyin commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+package duke;
+
+public class DukeException extends Exception {
```
Perhaps you can add some class header comments?

rockman007372 replied:
Added in the most recent commit!

lhy-hoyin commented on src/test/java/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,31 @@
+
+public class TaskListTest {
+    @Test
+    public void getTaskTest() {
```
Perhaps name the test method using the `whatIsBeingTested_descriptionOfTestInputs_expectedOutcome` convention?

lhy-hoyin commented on src/test/java/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,31 @@
+    @Test
+    public void deleteTaskTest() throws DukeException {
+        Task testTask = new ToDo("test");
+        Task dontCare = new ToDo("");
```
Not sure if I like how the `Task` is being used in SUT.

rockman007372 replied:
Do you recommend I should make Task a STUB? I am not sure what you mean here. Thanks.

lhy-hoyin commented on src/test/java/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Test
+    public void deleteTaskTest() throws DukeException {
```
Perhaps name the test method using the whatIsBeingTested_descriptionOfTestInputs_expectedOutcome convention?

rockman007372 replied:
Yes noted!

lhy-hoyin commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,13 @@
+    @Test
+    public void dummyTest(){
+        assertEquals(2, 2);
+    }
```
Not quite sure if this `dummyTest()` brings any value to unit testing the code.

lhy-hoyin commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,13 @@
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
```
Perhaps the static imports should come first?

lhy-hoyin commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,137 @@
+package duke.ui;
+
+import duke.command.*;
```
I don't like how wildcard imports are used here.

lhy-hoyin commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,58 @@
+package duke.task;
+
+public class Task {
```
Is there any reason why an `Task` is not abstract?

lhy-hoyin commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,58 @@
+     * Returns the string representation of the task in the save file.
+     * @return Representation of the task in the save file.
+     */
+    public String toSavedString() {
+        return "";
+    }
```
Since this returns an empty string, I assume that it is expected for the child classes to always override it. Perhaps make this an abstract function?

lhy-hoyin commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,137 @@
+
+        // Split strings into 2, first part is the operation, 2nd part is the description
+        String[] command = input.split(" ", 2);
+        Operation op = Operation.valueOf(command[0].toUpperCase());
```
I like how `valueOf` is being used here.

lhy-hoyin commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,137 @@
+            return findTaskParser(input);
+        }
+
+        return null; // cannot reach here, as duke.command.Operation.valueOf throws IllegalArgumentException
```
Perhaps put this inside the switch case as the `default` branch?

lhy-hoyin commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,137 @@
+            return new AddEventCommand(eventDescription, from, to);
+        }
+
+        return null;
```
Perhaps put this inside the switch case as the default branch?

lhy-hoyin commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,137 @@
+            return new AddTodoCommand(description);
+        case DEADLINE:
+            String[] deadlineString = description.split("/by", 2);
+            String deadlineDescription = deadlineString[0];
```
Perhaps `trim()` the description since there might be a `' '` before the `/by`?

rockman007372 replied:
This makes sense, thank you!

lhy-hoyin commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,172 @@
+                    foundTasks.append(System.lineSeparator());
+                }
+            }
+
+
```
Blank lines are not being used.

lhy-hoyin reviewed (commented):
Overall code quality seems good, although it seems that generally head comments are missing for the classes (as required by the coding standards).
I like how the README is update even though its not part of the requirement. However, it confused me as the seemingly unnecessary files were committed to git/github. For example, data files (i.e. `duke.txt`) and mysterious log files (i.e. `q`).

rockman007372 committed:
Move catch statements from Duke to Parser

rockman007372 committed:
Add hashmap to map index to task

rockman007372 committed:
Fix all checkstyle errors

rockman007372 committed:
Merge branch 'master' into branch-Level-10-JavaFX

Merge the javadoc from master.

rockman007372 committed:
Update JavaFX to Level 3

rockman007372 committed:
Update gitignore to ignore /data/

rockman007372 committed:
Change class names in duke.command package, and add trim() method

rockman007372 committed:
Refractor Parser and add CommandMark to combine mark and unmark operation

rockman007372 committed:
Modify deleteTaskTest so the test passes

rockman007372 committed:
Do complete overhaul of bot response UI

rockman007372 committed:
Add and modify test cases in TaskListTest

rockman007372 committed:
Add an echoing GUI

rockman007372 committed:
Add a functional but ugly GUI to Duke

rockman007372 committed:
Merge with branch Level 10 to add GUI

rockman007372 committed:
Fix bug where the tast index in the hashmap is not deleted

rockman007372 committed:
Add varargs to search using a list of keywords

rockman007372 committed:
Fix checkstyle

jishnu28 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+    private static final String FILE_DIRECTORY = "data";
+    private static final String FILE_NAME = "duke.txt";
+
+    private Storage storage;
```
I like the clear and simple naming of variables here.

jishnu28 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,19 @@
+        super(message);
+    }
+
+    /**
```
I think the header comment here can be omitted since its a simple overriding method

jishnu28 commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,12 @@
+
+import javafx.application.Application;
+
+/**
```
I think this is an interesting approach but I'm curious what classpath issues you might have run into

jishnu28 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Represents a command from user input.
+ */
+public abstract class Command {
```
I think its a good call to make this class abstract and to keep it simple. Good job!

jishnu28 commented on src/main/java/duke/command/Operation.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Represents a valid operation in the user input.
+ */
+public enum Operation {
```
I like the use of enums here.

jishnu28 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+package duke.task;
```
Is it possible for the TaskList class to be outside of the 'task' package? I was wondering this since it felt a bit odd that it's not similar to the other classes here but is related to tasks

rockman007372 replied:
I think this makes sense! (I was just following Ai Bo)

jishnu28 reviewed (commented):
LGTM. Interesting approach for some problems and good package structure.

rockman007372 committed:
Fix some code quality issues

rockman007372 committed:
Fix index out of bound bug when deleting new tast after deleting older ones

rockman007372 committed:
Fix bug where saying bye_ does not close program

rockman007372 committed:
Add a new jar file

rockman007372 committed:
Add some assertions and modify storage class

rockman007372 committed:
Edit README

rockman007372 committed:
Update .gitignore file

rockman007372 committed:
Remove CONTRIBUTORS.MD from tracked files

rockman007372 committed:
Remove .\data\ from tracked directory

rockman007372 committed:
Edit gitignore

rockman007372 committed:
Remove text-ui-test from tracked directory

rockman007372 committed:
Fix some code quality

rockman007372 committed:
Replace loops with stream and fix bug where the marked task shown is incorrect

rockman007372 committed:
Merge pull request #4 from rockman007372/branch-A-Streams

Replace loops with stream and fix bug in MarkTaskCommand

rockman007372 committed:
Remove data directory from tracked directory

rockman007372 committed:
Merge with master

rockman007372 committed:
Merge pull request #3 from rockman007372/branch-A-CodeQuality

Improve code quality

rockman007372 committed:
Merge with master

rockman007372 committed:
Merge pull request #2 from rockman007372/branch-A-Assertions

Add some assertions and modify storage class

rockman007372 committed:
Add assertions

rockman007372 committed:
Fix code quality and follow code convention

rockman007372 committed:
Change some loop iterations to streams

rockman007372 committed:
Add update functionality

rockman007372 committed:
Merge branch 'branch-C-Update'

rockman007372 committed:
Add update functionaity to the bot

Users cannot update the task directly, and must delete and add a new task as a hack.

Let's,
* create an update command
* allow users to modify a given task's components using command /description, /by, ...

Updating task directly saves users from having to type in repeated information when they create a new updated task.

rockman007372 committed:
Modify some UI settings

rockman007372 committed:
Enhance the parsing of Update command

rockman007372 committed:
Add a different colour to opposing dialog

rockman007372 committed:
Edit fxml files to prevent warnings

rockman007372 committed:
Merge branch 'A-BetterGui'

rockman007372 committed:
Modify welcome message

rockman007372 committed:
Handle more exceptions, mostly done in Parser class

rockman007372 committed:
A bit more testing

rockman007372 committed:
Fix checkstyle

rockman007372 committed:
Add Ui.png

rockman007372 committed:
Add User Guide

rockman007372 committed:
Modify user guide

rockman007372 committed:
Change some error messages

rockman007372 committed:
Remove print statement
